community/index/src/main/java/org/neo4j/index/internal/gbptree/CrashGenCleaner.java

```diff
@@ -0,0 +1,245 @@
+        if ( finalError != null )
+        {
+            throw launderedException( IOException.class, finalError );
+        }
+
+        long endTime = currentTimeMillis();
+        monitor.recoveryCompleted( highTreeNodeId - lowTreeNodeId, cleanedPointers.get(), endTime - startTime );
+    }
+
+    private Runnable cleaner( AtomicLong nextId, AtomicReference<Throwable> error, AtomicInteger cleanedPointers )
+    {
+        return () ->
+        {
+            try ( PageCursor cursor = pagedFile.io( 0, PagedFile.PF_SHARED_READ_LOCK );
+                    PageCursor writeCursor = pagedFile.io( 0, PagedFile.PF_SHARED_WRITE_LOCK ) )
+            {
+                while ( nextId.get() < highTreeNodeId )
+                {
+                    // Do a batch of max 1000
+                    int batchCount = 0;
+                    for ( ; batchCount < 1_000; batchCount++ )
+                    {
+                        long id = nextId.getAndIncrement();
+                        if ( id >= highTreeNodeId )
+                        {
+                            break;
+                        }
+
+                        // TODO: we should try to figure out if the disk backing this paged file
+                        // is good at concurrent random reads (i.e. if it's an SSD).
+                        // If it's not it'd likely be beneficial to let other threads wait until
+                        // goTo completes before letting others read. This will have reads be
+                        // 100% sequential.
+                        PageCursorUtil.goTo( cursor, "clean", id );
+
+                        if ( hasCrashedGSPP( treeNode, cursor ) )
+                        {
+                            writeCursor.next( cursor.getCurrentPageId() );
+                            cleanTreeNode( treeNode, writeCursor, cleanedPointers );
+                        }
+                    }
+
+                    // Check error status after a batch, to reduce volatility overhead.
+                    // Is this over thinking things? Perhaps
+                    if ( error.get() != null )
+                    {
+                        break;
+                    }
+                }
+            }
+            catch ( Throwable e )
+            {
+                error.compareAndSet( null, e );
+            }
+        };
+    }
+
+    // === Methods about checking if a tree node has crashed pointers ===
+
+    private boolean hasCrashedGSPP( TreeNode<?,?> treeNode, PageCursor cursor ) throws IOException
+    {
+        boolean isTreeNode;
+        int keyCount;
+        do
+        {
+            isTreeNode = TreeNode.nodeType( cursor ) == TreeNode.NODE_TYPE_TREE_NODE;
+            keyCount = treeNode.keyCount( cursor );
+        }
+        while ( cursor.shouldRetry() );
+
+        if ( !isTreeNode )
+        {
+            return false;
+        }
+
+        boolean hasCrashed;
+        do
+        {
+            hasCrashed =
```

Comment:
`hasCrashedGSPP` takes exact offset as argument and given offsets are absolute so we don't rely on position from previous call when progressing with next read.